### DISCUSSION TITLE
Replacing Luasnip with Ultisnips

### DISCUSSION THREAD

**Discussion by firubat** [LazyVim/LazyVim#1474] [Q&A] [answered]
I'm trying to replace Luasnip with Ultisnip (mainly because I don't want to rewrite all my snippets...). I tried adding a snippets.lua file in the plugins folder, and replacing LazyVim's config of nvim-cmp with a modified one, based on [nvim-cmp docs:](https://github.com/hrsh7th/nvim-cmp/tree/main#recommended-configuration)

```
return {
  { "L3MON4D3/LuaSnip", enabled = false }, -- disable luasnip
  {
    "SirVer/ultisnips",
    init = function()
      vim.g.UltiSnipsSnippetDirectories = { "~/.config/nvim/UltiSnips" }
    end,
  }, -- enable ultisnip
  { "quangnguyen30192/cmp-nvim-ultisnips" }, --integrate ultisnips with nvim-cmp for completion
  {
    "hrsh7th/nvim-cmp",
    version = false, -- last release is way too old
    event = "InsertEnter",
    dependencies = {
      "hrsh7th/cmp-nvim-lsp",
      "hrsh7th/cmp-buffer",
      "hrsh7th/cmp-path",
      "quangnguyen30192/cmp-nvim-ultisnips",
      --"saadparwaiz1/cmp_luasnip",
    },
    opts = function()
      vim.api.nvim_set_hl(0, "CmpGhostText", { link = "Comment", default = true })
      local cmp = require("cmp")
      local defaults = require("cmp.config.default")()
      return {
        completion = {
          completeopt = "menu,menuone,noinsert",
        },
        snippet = {
          expand = function(args)
            vim.fn["UltiSnips#Anon"](args.body) -- For `ultisnips` users.
            --require("luasnip").lsp_expand(args.body)
          end,
        },
        mapping = cmp.mapping.preset.insert({
          ["<C-n>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Insert }),
          ["<C-p>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Insert }),
          ["<C-b>"] = cmp.mapping.scroll_docs(-4),
          ["<C-f>"] = cmp.mapping.scroll_docs(4),
          ["<C-Space>"] = cmp.mapping.complete(),
          ["<C-e>"] = cmp.mapping.abort(),
          ["<CR>"] = cmp.mapping.confirm({ select = true }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
          ["<S-CR>"] = cmp.mapping.confirm({
            behavior = cmp.ConfirmBehavior.Replace,
            select = true,
          }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
        }),
        sources = cmp.config.sources({
          { name = "nvim_lsp" },
          --{ name = "luasnip" },
          { name = "ultisnips" },
          { name = "buffer" },
          { name = "path" },
        }),
        formatting = {
          format = function(_, item)
            local icons = require("lazyvim.config").icons.kinds
            if icons[item.kind] then
              item.kind = icons[item.kind] .. item.kind
            end
            return item
          end,
        },
        experimental = {
          ghost_text = {
            hl_group = "CmpGhostText",
          },
        },
        sorting = defaults.sorting,
      }
    end,
  },
}
```

But when I try to load it I get the following error message:

```
Failed loading config.options

vim/_meta.lua:0: Invalid 'option': 'vim'

# stacktrace:
  - vim/_meta.lua:0 _in_ **__index**
  - ~/.config/nvim/lua/config/options.lua:10
  - /LazyVim/lua/lazyvim/config/init.lua:143
  - /LazyVim/lua/lazyvim/config/init.lua:142 _in_ **_load**
  - /LazyVim/lua/lazyvim/config/init.lua:159 _in_ **load**
  - /LazyVim/lua/lazyvim/config/init.lua:178 _in_ **init**
  - /LazyVim/lua/lazyvim/plugins/core.lua:1
  - ~/.config/nvim/lua/config/lazy.lua:9
  - ~/.config/nvim/init.lua:2

```

Any help appreciated!

**Comment by dpetka2001** [comment, marked as answer]
Could you try the following
```lua
return {
  {
    "L3MON4D3/LuaSnip",
    enabled = false,
  },

  {
    "SirVer/ultisnips",
    dependencies = {
      {
        "honza/vim-snippets",
      },
    },
    -- init = function()
      -- vim.g.UltiSnipsSnippetDirectories = { "~/.config/nvim/UltiSnips" }
    -- end,
  },

  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "quangnguyen30192/cmp-nvim-ultisnips",
      { "saadparwaiz1/cmp_luasnip", enabled = false },
    },
    opts = function(_, opts)
      local cmp = require("cmp")
      opts.snippet = {
        expand = function(args)
          vim.fn["UltiSnips#Anon"](args.body) -- For `ultisnips` users.
        end,
      }
      opts.sources = cmp.config.sources({
        { name = "nvim_lsp" },
        --{ name = "luasnip" },
        { name = "ultisnips" },
        { name = "buffer" },
        { name = "path" },
      })
    end,
  },
}
```
I was not able to remove `cmp-luasnip` as a dependency of `nvim-cmp` (I tried passing a `values` function to overwrite the whole `dependencies` default section of `nvim-cmp` but I got an error about not being able to get the total length of the spec and after looking at the `handler` code in `lazy.nvim`, I did not see `dependencies` listed as handlers so maybe this why I can't pass a `values` function to overwrite it?), so if anyone else has a more complete solution please come forward.

**Reply by abeldekat** [reply]
Look's good, @dpetka2001!

The dependencies property only `extends` across multiple specs for the same plugin. See [plugin.lua](https://github.com/folke/lazy.nvim/blob/2a9354c7d2368d78cbd5575a51a2af5bd8a6ad01/lua/lazy/core/plugin.lua#L379)

Internally, a bit simplified, when *parsing* the spec, all specs inside the dependencies of the spec are processed as individual plugin specs, and added to the internal plugin list. 

The dependencies property of the resulting `object` of the parent plugin contains only the *names* of the specified plugins, as a list of strings.

When `lazy.nvim ` *loads* the plugin, its dependencies are loaded. Any plugin specified as a dependency  will not be loaded when that plugin is disabled. See [loader.lua](https://github.com/folke/lazy.nvim/blob/2a9354c7d2368d78cbd5575a51a2af5bd8a6ad01/lua/lazy/core/loader.lua#L176)

Thus, when you disable a plugin defined as a dependency, you effectively remove the dependency, although the name of that plugin is still present in the dependencies property of the parent plugin.

**Reply by dpetka2001** [reply]
As always your thorough explanation is greatly appreciated. Thank you for your time :smile:

**Reply by firubat** [reply]
@dpetka2001 Thanks, so now UltiSnips manages to load, but at every character I try to insert gives an error that looks like:

```
Following is the full stack trace:
Traceback (most recent call last):
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/err_to_scratch_buffer.py", line 47, in wrapper
    return func(self, *args, **kwds)
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/snippet_manager.py", line 945, in _track_change
    self._try_expand(autotrigger_only=True)
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/snippet_manager.py", line 802, in _try_expand
    before, snippets = self._can_expand(autotrigger_only)
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/snippet_manager.py", line 798, in _can_expand
    return before, self._snips(before, False, autotrigger_only)
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/snippet_manager.py", line 679, in _snips
    source.ensure(filetypes)
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/snippet/source/file/base.py", line 32, in ensure
    self._load_snippets_for(ft)
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/snippet/source/file/base.py", line 54, in _load_snippets_for
    self._parse_snippets(ft, fn)
  File "/home/ur/.local/share/nvim/lazy/ultisnips/pythonx/UltiSnips/snippet/source/file/base.py", line 62, in _parse_snippets
    with open(filename, "r", encoding="utf-8-sig") as to_read:
IsADirectoryError: [Errno 21] Is a directory: '/home/ur/.local/share/nvim/lazy/plenary.nvim/lua/luassert'
```

**Reply by dpetka2001** [reply]
Sorry, but I'm not familiar with Ultisnips. I didn't have this problem. I just used `honza/vim-snippets` as the snippet engine and tried it in a lua file while disabling Luasnip. I tried to type `for` and it showed a snippet about the `for` loop and accepted that. I also tried an `if...else` snippet which also succeeded. Didn't test anything more. I can't tell you what this error is about. Did you use the exact piece of code above or did you make any changes? Do you have your nvim dotfiles available on a github repo to take a look and see if I can find anything that might look suspicious or not? Also it might have something to do with the `nvim-cmp` itself or Neovim version even (I'm using Neovim built from source updated some time yesterday).  What file did you test it on? Can you test with other filetypes as well to be sure they all behave the same? Since the plugin loads it shouldn't be Lazyvim related any more but some kind of other misconfiguration I would guess? In that case your best bet would be to make an issue at Ultisnips repo and let the people who are more knowledgeable about the plugin help you, because as I've already mentioned I'm not familiar with it at all. I just showed you the Lazyvim way to install and load it, but have absolutely no idea about how it works.

**Reply by firubat** [reply]
@dpetka2001 OK OK, so it seems that there is some problem with my snippets, because if I remove them, and comment out the lines pointing to them and only use the snippets from honza/vim-snippets, things work out fine. So maybe remove/comment out those lines in your answer and I'll mark it as answer? I'm talking about these lines:
```
init = function()
      vim.g.UltiSnipsSnippetDirectories = { "~/.config/nvim/UltiSnips" }
    end,
```
which basically tells UltiSnips to use only the snippets in ~/.config/nvim/UltiSnips. I think also the ~ doesn't expand well.

And I'll try to figure out which of my snippets break things...

**Reply by dpetka2001** [reply]
@firubat I commented out the lines you mentioned. I only included those lines because you had them in your initial post, so I didn't want to alter the location of your snippets.

**Reply by sinkr** [reply]
Holy Mackerel @dpetka2001!  I lost two days trying to get what you have here working!  Thank you so much!  I really need to learn Lua apparently.
